include/dxc/DXIL/DxilFunctionProps.h

```diff
@@ -25,6 +25,93 @@ class Constant;
 } // namespace llvm
 
 namespace hlsl {
+
+// SM 6.6 allows WaveSize specification for only a single required size.
+// SM 6.8+ allows specification of WaveSize as a min, max and preferred value.
+struct DxilWaveSize {
+  unsigned Min = 0;
+  unsigned Max = 0;
+  unsigned Preferred = 0;
+
+  DxilWaveSize() = default;
+  DxilWaveSize(unsigned min, unsigned max = 0, unsigned preferred = 0)
+      : Min(min), Max(max), Preferred(preferred) {}
+  DxilWaveSize(const DxilWaveSize &other) = default;
+  DxilWaveSize &operator=(const DxilWaveSize &other) = default;
+  bool operator==(const DxilWaveSize &other) const {
+    return Min == other.Min && Max == other.Max && Preferred == other.Preferred;
+  }
+
+  // Encode valid DxilWaveSize, handling potential degenerate cases.
+  static DxilWaveSize Encode(unsigned min, unsigned max = 0,
+                             unsigned preferred = 0) {
+    if (max == min)
+      max = 0;
```

Comment:
Per the spec, max values aren't allowed to be zero and min must be less than max. This turns one invalid state to another.

Reply:
> This turns one invalid state to another.

No, it doesn't.  It translates the degenerate cases, which are valid cases, but aren't represented that way in this structure or in metadata.  Truly invalid cases will still result in invalid state, which is validated elsewhere.
Min == Max is the meaning of max = 0, and the correct encoding for PSV0 and RDAT.  Missing this was the reason for the regression (assuming that Max = 0 was correct for PSV0).

And I don't see how this is translating into an invalid state.

You seem to be thinking about this struct as if it *is* the metadata.  It isn't.  It is the internal representation for any possible state for WaveSize.